src/custom-apps/components/PermissionAlert/utils.ts

```diff
@@ -50,92 +56,119 @@ const uniq = <T>(value: T, index: number, self: T[]) => self.indexOf(value) ===
+  }, [] as GraphQLObjectType[]);
+
+const getSubscriptions = (typeMap: TypeMap): GraphQLObjectType[] =>
+  Object.keys(typeMap).reduce((acc, key) => {
+    const type = typeMap[key] as GraphQLObjectType;
+
+    if (type instanceof GraphQLObjectType) {
+      const interfaces = type.getInterfaces();
+      const hasEvent = interfaces.some(({ name }) => name === "Event");
 
-    visit(ast, {
-      Field(node, _key, _parent, _path, ancestors) {
-        if (node.name.value !== "__typename") {
-          const cursor = ancestors.filter(byKind("Field")).map(extractValue).filter(isNotEvent);
+      if (hasEvent) {
+        return [...acc, type];
+      }
+    }
+
+    return acc;
+  }, [] as GraphQLObjectType[]);
+
+function getDescriptionsFromQuery(query: string, schema: GraphQLSchema) {
+  const descriptions: { [key: string]: string } = {};
+  const ast = parse(query);
+  const startToken = getStartToken(ast);
+  const tree = tokenTree(startToken, []);
+  const subscriptions = getSubscriptions(schema.getTypeMap());
+  const subscriptionsFromQuery = extractSubscriptions(subscriptions, tree);
+
+  visit(ast, {
+    Field(node, _key, _parent, _path, ancestors) {
+      for (const _type in subscriptionsFromQuery) {
+        const fieldPath = ancestors
+          .filter((ancestor: any) => ancestor.kind === "Field")
+          .map((ancestor: any) => ancestor.name.value)
+          .concat(node.name.value);
+        let type: GraphQLObjectType | undefined = subscriptionsFromQuery[_type];
+
+        for (const fieldName of fieldPath.slice(1, -1)) {
+          if (type) {
+            const field: GraphQLField<any, any, any> = (
+              unwrapType(type) as GraphQLObjectType
+            ).getFields()[fieldName];
+
+            // Some types can have multiple nested fields eg. lists
+            type = field && field.type ? (unwrapType(field.type) as GraphQLObjectType) : undefined;
+          }
+        }
 
-          if (cursor.length > 0) {
-            cursors.push([...cursor, node.name.value]);
+        if (type) {
+          const field = type.getFields()[node.name.value];
+
+          if (field) {
+            descriptions[fieldPath.join(".")] = field.description || "No description available";
           }
         }
-      },
-    });
+      }
+    },
+  });
+
+  return descriptions;
+}
+
+const extractPermissionsFromQuery = (
+  query: string,
+  introspectionQuery: IntrospectionQuery,
+): SubscriptionQueryPermission => {
+  let permissions: SubscriptionQueryPermission = {};
+
+  try {
+    const schema = buildClientSchema(introspectionQuery);
+    const descriptions = getDescriptionsFromQuery(query, schema);
+
+    const _permissions = Object.keys(descriptions).reduce((acc, key) => {
+      const { isOneOfRequired, permissions } = extractPermissions(descriptions[key]);
+
+      if (permissions.length > 0) {
+        acc[key] = { isOneOfRequired, permissions };
+      }
+
+      return acc;
+    }, {} as SubscriptionQueryPermission);
```

Comment:
suggestion: That could be also be wrapped into separate function and `extractPermissionsFromQuery ` would look like this:

```suggestion
const extractPermissionsFromQuery = (
  query: string,
  introspectionQuery: IntrospectionQuery,
): SubscriptionQueryPermission => {
    try {
      const schema = buildClientSchema(introspectionQuery);
      const descriptions = getDescriptionsFromQuery(query, schema);
      
      return getAllPermission(descriptions) // You name it better 
    } catch (error) {
      // explicit silent
    }
}
```